Workflow Editor UI for Step Labels

jmchilton opened the pull request:
About a [year ago](https://github.com/galaxyproject/galaxy/commit/7dfdf530cf3c902f6de99c0e7783fd73312043e6), workflow steps were augmented to uniformly allow labels. Labels differ from the concept of dataset input names in that they:
- Are enforced to be unique across a workflow, providing a preserved and human consumable way to reference steps across workflow saves.
- They are uniformly treated - tool steps for instance can have a label.

Step labels have been used extensively in the API, testing, and in the workflow format 2 specification but have not been exposed in the UI previously. My downstream subworkflow/CWL will make use of labels in the UI to describe inputs (along as extending workflow outputs with their own concept of labels), and will be required for display and distinguishing of subworkflows node titles (since subworkflows may or may not correspond to a StoredWorkflow which is the entity that has a name). So I believe this concept should be exposed in the UI for subworkflows to be maximally usable (not to mention format 2 worfklows, CWL workflows, etc...).

Even existing workflows could potentially be improved by allowing the workflow author to specify labels for inputs to and tools that appear multiple places in the the workflow to distinguish between the operations.

This PR does this and tackles displaying and distinguishing nodes more broadly. In particular:
- If a step has a label, display it in the step editor side panel title. e3161b4671e9fb19a47c7f341e3fb46617111a9e
- The editable text module is used to allow clicking the title to change the label. e3161b4671e9fb19a47c7f341e3fb46617111a9e
- Display the label (if set) as the workflow node box title. e3161b4671e9fb19a47c7f341e3fb46617111a9e 
- Establish font awesome icons for all step types and display in the editor and on the node (brings all modules inline with tools in the side editor, icon will be useful on the node because the label can not mask the type of node). d7ac52f6bb61fe360198f91310a3e22076183d7b
- Enforce unique labels across the workflow in the editor. e3161b4671e9fb19a47c7f341e3fb46617111a9e
- Bring tool and non-tool step editor styling in sync in other ways. d18c75ceb875632f9e68a74ebeb88389262d6fa8
- Qunit test cases for some of this behavior.  e3161b4671e9fb19a47c7f341e3fb46617111a9e
- Various cleanups requested by @bgruening b11002b 2aa7f04 9824905
- Other small refactoring of existing stuff to make room for subworkflows f17a6c9 f6d26c9 73bfbf0 f92220a

The biggest problem with this PR I think is that it doesn't do anything to mask the concept of a dataset input name. While for backward compatibility Galaxy will continue to support names on inputs and labels as separate on the backend - the editor should modify the workflow when it loads to convert every name into a label (mangling if needed to make them unique), warn the user that this will be made permanent when it is saved, and then hide the name field from dataset inputs. From that point forward, the editor should just keep names and labels in sync and only expose labels to the user. I guess the question is whether this should happen before this is merged or if it can wait a release.

Another big TODO is to allow editing labels on tool steps, I have just been modifying non-tool modules since these modules are more important for subworkflows and the code is local to the workflow stuff instead of reused with the tool form.

![editor](https://cloud.githubusercontent.com/assets/216771/11614997/10d0e9ce-9c4c-11e5-9e9c-c811212aab9a.png)

jmchilton committed:
Small refactor making workflow views and modules tad more generic.

 - Reduce duplication between initializing generic modules and tool modules.
 - Switch tool_id to content_id as variable names throughout the client.
 - Rename get_tool_id to get_content_id on workflow modules.
 - Add some minimal documentation to the workflow module about get_content_id.

Downstream in the subworkflow commit I switch the over-the-wire communication to use content_id instead of tool_id also and use content_ids to refer to workflow ids in subworkflow moduls.

jmchilton committed:
Refactor some workflow API code into workflows manager.

This allows for reuse outside of the API controller, such as loading subworkflows downstream.

jmchilton committed:
Synchronize decoding ids across base controller and manager base.

To ensure we have the same kind of exception thrown in both cases.

jmchilton committed:
Allow workflow outputs on non-tool steps.

jmchilton committed:
Don't iterate over unused key.

Thanks to @bgruening.

jmchilton committed:
Replace __rip_value with dict.pop.

Thanks @bgruening.

jmchilton committed:
Cleanup error message.

Thanks @bgruening.

jmchilton committed:
Bring gneric and tool module styles into alignment in workflow editor.

jmchilton committed:
Add icons to existing workflow steps in workflow editor.

Tools have an icon there.

jmchilton committed:
Add support for workflow step labels in editor.

 - If a step has a label, display it in the step editor side panel title.
 - Allow clicking the title to change the label.
 - Display the label (if set) as the workflow node box title.
 - Add icon to workflow node since the title might not be related to type anymore.
 - Enforce unique labels accross the workflow in the editor.
 - Qunit test cases for some of this behavior and other recent changes.

bgruening commented:
Anyone wants to play with it? You can simply run:

```
docker run -p 8080:80 -e GALAXY_CONFIG_ENABLE_BETA_WORKFLOW_MODULES=True -i -t bgruening/galaxy-stable:jmc_workflow
```

nsoranzo commented:
LGTM

martenson commented:
Please do not merge, I have an addition.

martenson committed:
show notifications as Toasts and not like JS alerts (bleh)

martenson commented:
@jmchilton please review https://github.com/jmchilton/galaxy/pull/12

martenson commented:
Regarding renaming of the node - this field behaves differently than the others. E.g. if you type in the tag and click outside of the element, it will save it. It won't for the node. This inconsistency feels confusing.

martenson commented on client/galaxy/scripts/mvc/workflow/workflow-manager.js:
```diff
@@ -7,10 +7,49 @@ define(['mvc/workflow/workflow-connector'], function( Connector ) {
-            var node = this.app.prebuildNode( type, title_text, tool_id );
+        canLabelNodeWith: function( label ) {
+            if( label ) {
+                return ! (label in this.nodeLabels);
```
This expression feels pythonic. But that is matter of taste I guess as I do not see a clearly superior alternative. I just did not see this before.

jmchilton replied:
I'll take an unobviously superior alternative if it feels more JavaScript-y, I don't do much JavaScript so I am happy to take feedback - even purely stylistic things.

For this particular set of functions and attribute I wanted a set which is only supported in EMCA 6, so I used the top voted pattern from [this StackOverflow article](http://stackoverflow.com/questions/7958292/mimicking-sets-in-javascript).

martenson replied:
I would use the following, which is very easy to read for me. The choice is yours.
`return ( label in this.nodeLabels ) ? false : true;`

guerler replied:
Yet another similar alternative might be: `return !this.nodeLabels[ label ];`

jmchilton committed:
Merge pull request #12 from martenson/labels_toastr

show notifications as Toasts and not like JS alerts (bleh)

jmchilton committed:
Workflow label UI refinement.

 - Escape workflow labels in mako.
 - Handle unlabelling steps properly (restore original node name to node and editor form titles).

jmchilton commented:
Pushed a few updates to this:
- @martenson awesome work to improve alerts.
- Escaped workflow label in the mako. (Is there more I need to do in the jquery?).
- Handled "unlabelling" steps and restoring the original title to the editor title and node.

jmchilton commented:
@martenson When you say tag do you mean dataset tags in the history panel?

martenson commented:
@jmchilton I mean these:
![screenshot 2015-12-08 11 18 10](https://cloud.githubusercontent.com/assets/1814954/11660930/65ff1e32-9d9d-11e5-8d8b-626e963ed955.png)
sorry for the naming confusion

jmchilton commented:
@martenson I don't know how to solve this really -I would just say without data that I do feel like people are used to editable text elements being different from other form elements in this respect. I am open to ideas though.

martenson commented:
@jmchilton Fair enough, History name changes behave the same way.

jmchilton commented:
Awesome sauce, thanks for the merge.

nsoranzo commented:
This is breaking the TS tests in `test/tool_shed/functional/test_0060_workflows.py` , see https://jenkins.galaxyproject.org/job/docker-toolshed/21/testReport/ .
